### PR TITLE
feat(swc): Allow custom configuration of SWC CLI options in `nest-cli.json`

### PR DESCRIPTION
**Pull request by medz**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #2138  #2139 

## What is the new behavior?

- Fix SWC builder not set `sourceRoot` to SWC filenames
- feat: Allow custom output dir readable from `tsconfig.json`

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: lib/compiler/defaults/swc-defaults.ts [hunk 1]
```diff
@@ -25,8 +29,8 @@ export const swcDefaultsFactory = (tsOptions: ts.CompilerOptions) => {
       swcrc: true,
     },
     cliOptions: {
-      outDir: 'dist',
-      filenames: ['src'],
+      outDir: tsOptions.outDir ?? 'dist',
+      filenames: [configuration.sourceRoot ?? 'src'],
```
Comment (kamilmysliwiec) [inline lib/compiler/defaults/swc-defaults.ts:33]: "sourceRoot" is dedicated to schematics - it's not supposed to control builder's behavior. I think we should just expose the "swcCliOptions" prop (under the `compilerOptions` in `nest-cli.json`) that would let everyone override the default options - if needed.

Would you like to update the PR?

Comment (medz) [reply lib/compiler/defaults/swc-defaults.ts:33]: Ok, I will update him, but this change will involve a schema change in `nest-cli.json`, also, the default outDir is set to `outDir` in tsconfig, which is more in line with configuration expectations.

**Comment by medz** [comment]
@kamilmysliwiec 
I've done your suggestion and can now configure swc cli in `nest-cli.json`:

```json
{
  "collection": "@nestjs/schematics",
  "sourceRoot": "server",
  "compilerOptions": {
    "deleteOutDir": true,
    "builder": "swc",
    "swcCliOptions": {
      "filenames": ["server"],
      // ... and more
    }
  }
}
```

All options see https://github.com/nestjs/nest-cli/pull/2140/commits/3b1c1c4c0d3c8511e8ce435dcf4dc64051ed6d1e#diff-463cedca96ba780a17deab2df14ba681ba9f6f25f76c43923de87835b08902eeR19

File: lib/compiler/defaults/swc-defaults.ts [hunk 2]
```diff
@@ -25,14 +29,14 @@ export const swcDefaultsFactory = (tsOptions: ts.CompilerOptions) => {
...
+      filenames: swcCliOptions?.filenames || ['src'],
+      sync: swcCliOptions?.sync || false,
+      extensions: swcCliOptions?.extensions || ['.js', '.ts'],
       watch: false,
-      copyFiles: false,
-      includeDotfiles: false,
-      quiet: false,
+      copyFiles: swcCliOptions?.copyFiles || false,
+      includeDotfiles: swcCliOptions?.includeDotfiles || false,
+      quiet: swcCliOptions?.quiet || false,
```
Comment (kamilmysliwiec) [inline lib/compiler/defaults/swc-defaults.ts:39]: Wouldn't it be easier to do `...swcCliOptions` instead?

File: lib/configuration/configuration.ts [hunk 3]
```diff
@@ -16,6 +16,16 @@ export interface ActionOnFile {
   watchAssetsMode: boolean;
 }
 
+export interface SwcCliOptions {
```
Comment (kamilmysliwiec) [inline lib/configuration/configuration.ts:19]:
```suggestion
interface SwcCliOptions {
```

File: lib/compiler/defaults/swc-defaults.ts (hunk 2 above)
Comment (medz) [reply lib/compiler/defaults/swc-defaults.ts]: Indeed destructuring is easier to maintain defaults, changed and committed

File: lib/configuration/configuration.ts (hunk 3 above)
Comment (medz) [reply lib/configuration/configuration.ts]: Export of `SwcCliOptions` has been removed

File: lib/compiler/defaults/swc-defaults.ts [hunk 4]
```diff
@@ -25,14 +33,17 @@ export const swcDefaultsFactory = (tsOptions: ts.CompilerOptions) => {
...
       filenames: ['src'],
       sync: false,
       extensions: ['.js', '.ts'],
-      watch: false,
       copyFiles: false,
       includeDotfiles: false,
       quiet: false,
+      watch: false,
+
+      // Merge swcCliOptions
```
Comment (kamilmysliwiec) [inline lib/compiler/defaults/swc-defaults.ts:45]:
```suggestion
```

File: lib/compiler/defaults/swc-defaults.ts [hunk 5]
```diff
@@ -25,14 +33,17 @@ export const swcDefaultsFactory = (tsOptions: ts.CompilerOptions) => {
       swcrc: true,
     },
     cliOptions: {
-      outDir: 'dist',
+      outDir: tsOptions.outDir || 'dist',
```
Comment (kamilmysliwiec) [inline lib/compiler/defaults/swc-defaults.ts:36]:
```suggestion
      outDir: tsOptions.outDir ?? 'dist',
```

File: lib/compiler/defaults/swc-defaults.ts (hunk 4 above)
Comment (medz) [reply lib/compiler/defaults/swc-defaults.ts]: Removed it.

**Comment by medz** [comment]
@kamilmysliwiec Is there a plan for when it will be implemented?

**Comment by kamilmysliwiec** [comment]
Thank you @medz! I made a few changes and so the API now looks as follows:

```json
{
  "$schema": "https://json.schemastore.org/nest-cli",
  "collection": "@nestjs/schematics",
  "sourceRoot": "src",
  "compilerOptions": {
    "deleteOutDir": true,
    "typeCheck": true,
    "builder": {
      "type": "swc",
      "options": {
        "filenames": ["src", "test"],
      }
    }
  }
}
```

If you don't need to pass custom options, then you can just set `builder` to `"swc"` (so the original syntax is still supported)

**Comment by medz** [comment]
@kamilmysliwiec Awesome, I've been waiting for nest + swc for a long time ~ thank you for making such a great framework

**Pull request merged**